crates/jstz_core/src/kv/transaction.rs

```diff
@@ -105,86 +105,65 @@ impl SnapshotEntry {
-                    if !entry.dirty {
-                        Some(k.clone())
-                    } else {
-                        None
-                    }
-                },
-            )
-            .collect()
-    }
-
-    fn insert_set(&self) -> BTreeSet<OwnedPath> {
-        self.snapshot
-            .iter()
-            .filter_map(
-                |(k, entry)| {
-                    if entry.dirty {
-                        Some(k.clone())
-                    } else {
-                        None
-                    }
-                },
-            )
-            .collect()
-    }
-
-    pub(crate) fn update_set(&self) -> BTreeSet<OwnedPath> {
-        self.insert_set().union(&self.remove_set).cloned().collect()
-    }
-
-    pub(crate) fn flush(self, rt: &mut impl Runtime) -> Result<()> {
-        // Perform deletions
-        for key in self.remove_set {
-            Storage::remove(rt, &key)?
-        }
-
-        // Perform insertions
-        for (key, entry) in self.snapshot.into_iter() {
-            if entry.dirty {
-                Storage::insert(rt, &key, entry.value.as_ref())?;
-            }
+impl Transaction {
+    pub fn new() -> Self {
+        Self {
+            parent: None,
+            remove_set: BTreeSet::new(),
+            snapshot: BTreeMap::new(),
         }
-
-        Ok(())
     }
 
     fn lookup<'a, V>(
         &'a mut self,
         rt: &impl Runtime,
         key: OwnedPath,
-    ) -> Result<Option<&'a mut SnapshotEntry>>
+    ) -> Result<Option<&mut SnapshotEntry>>
     where
         V: Value + DeserializeOwned,
     {
-        let entry = self.snapshot.entry(key);
-
-        match entry {
-            btree_map::Entry::Vacant(entry) => {
-                if let Some(value) = Storage::get::<V>(rt, entry.key())? {
-                    let snapshot_entry = entry.insert(SnapshotEntry::persistent(value));
-
-                    return Ok(Some(snapshot_entry));
+        let first_entry = self.snapshot.entry(key.clone());
+
+        // Recursively lookup in parent if not found in current snapshot. If found in parent, insert into current snapshot.
+        match first_entry {
+            btree_map::Entry::Vacant(entry) => match &self.parent {
+                Some(parent) => {
+                    let parent = &mut parent.deref().borrow_mut();
+                    let parent_entry = parent.lookup::<V>(rt, key.clone())?;
+                    match parent_entry {
+                        Some(value) => {
+                            let snapshot_entry = entry.insert(value.clone());
+                            return Ok(Some(snapshot_entry));
+                        }
+                        None => {
+                            return Ok(None);
+                        }
+                    }
                 }
+                None => {
+                    if Storage::contains_key(rt, entry.key())? {
+                        let value = Storage::get::<V>(rt, entry.key())?.unwrap();
```

Comment:
Can't we simply use `Storage::get` here?

Reply:
Additionally would avoid an `.unwrap()`

Reply:
I removed the unwrap. Storage::get requires type annotations.